File: platforms/ios/Tcall/Services/ApiClient.swift
import Foundation

@MainActor
final class ApiClient {
    static let shared = ApiClient()
    private let session: URLSession
    private(set) var token: String?

    private init() {
        session = URLSession(configuration: .default)
    }

    private func request(_ path: String, method: String = "GET", body: Data? = nil) throws -> URLRequest {
        var req = URLRequest(url: TcallConfig.apiBaseURL.appendingPathComponent(path))
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(TcallConfig.nativeHeaderValue, forHTTPHeaderField: TcallConfig.nativeHeader)
        if let token {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody = body
        return req
    }

    func login(email: String, password: String) async throws -> UserDto {
        let payload = LoginRequest(email: email.lowercased().trimmingCharacters(in: .whitespaces), password: password, remember: true)
        let data = try JSONEncoder().encode(payload)
        let req = try request("/api/auth/login", method: "POST", body: data)
        let (resData, res) = try await session.data(for: req)
        guard let http = res as? HTTPURLResponse else { throw ApiError.network }
        let decoded = try JSONDecoder().decode(LoginResponse.self, from: resData)
        if http.statusCode == 401 { throw ApiError.message(decoded.error ?? "Email yoki parol noto'g'ri") }
        guard let user = decoded.user, let t = decoded.token, !t.isEmpty else {
            throw ApiError.message(decoded.error ?? "Kirish xatosi")
        }
        token = t
        KeychainStore.save(token: t)
        return user
    }

    func restoreSession() async throws -> UserDto? {
        guard let t = KeychainStore.loadToken() else { return nil }
        token = t
        let req = try request("/api/auth/session")
        let (data, res) = try await session.data(for: req)
        guard let http = res as? HTTPURLResponse, http.statusCode == 200 else {
            KeychainStore.clear()
            token = nil
            return nil
        }
        let decoded = try JSONDecoder().decode(SessionResponse.self, from: data)
        return decoded.user
    }

    func logout() async {
        var req = (try? request("/api/auth/session", method: "DELETE")) ?? URLRequest(url: TcallConfig.apiBaseURL)
        req.httpMethod = "DELETE"
        _ = try? await session.data(for: req)
        token = nil
        KeychainStore.clear()
    }
}

enum ApiError: LocalizedError {
    case network
    case message(String)
    var errorDescription: String? {
        switch self {
        case .network: return "Tarmoq xatosi"
        case .message(let s): return s
        }
    }
}

enum KeychainStore {
    private static let service = "uz.tcall"
    private static let account = "session"

    static func save(token: String) {
        let data = Data(token.utf8)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: data,
        ]
        SecItemDelete(query as CFDictionary)
        SecItemAdd(query as CFDictionary, nil)
    }

    static func loadToken() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data,
              let str = String(data: data, encoding: .utf8) else { return nil }
        return str
    }

    static func clear() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

File: packages/desktop-core/ApiClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json.Serialization;

namespace Tcall.Desktop.Core;

public sealed class UserDto
{
    [JsonPropertyName("userId")] public string UserId { get; set; } = "";
    [JsonPropertyName("email")] public string Email { get; set; } = "";
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("language")] public string Language { get; set; } = "uz";
    [JsonPropertyName("tcallId")] public string TcallId { get; set; } = "";
}

public sealed class LoginRequest
{
    [JsonPropertyName("email")] public string Email { get; set; } = "";
    [JsonPropertyName("password")] public string Password { get; set; } = "";
    [JsonPropertyName("remember")] public bool Remember { get; set; } = true;
}

public sealed class LoginResponse
{
    [JsonPropertyName("user")] public UserDto? User { get; set; }
    [JsonPropertyName("token")] public string? Token { get; set; }
    [JsonPropertyName("error")] public string? Error { get; set; }
}

public sealed class SessionResponse
{
    [JsonPropertyName("user")] public UserDto? User { get; set; }
}

public sealed class AuthSession
{
    public string Token { get; init; } = "";
    public UserDto User { get; init; } = new();
}

public sealed class ApiClient
{
    private readonly HttpClient _http;
    private string? _token;

    public ApiClient()
    {
        _http = new HttpClient { BaseAddress = new Uri(TcallConfig.ApiBaseUrl) };
        _http.DefaultRequestHeaders.Add(TcallConfig.NativeHeader, TcallConfig.NativeHeaderValue);
    }

    public UserDto? CurrentUser { get; private set; }

    public async Task<AuthSession> LoginAsync(string email, string password, CancellationToken ct = default)
    {
        var res = await _http.PostAsJsonAsync("/api/auth/login", new LoginRequest
        {
            Email = email.Trim().ToLowerInvariant(),
            Password = password,
        }, ct);

        var body = await res.Content.ReadFromJsonAsync<LoginResponse>(cancellationToken: ct);
        if (!res.IsSuccessStatusCode || body?.User is null || string.IsNullOrWhiteSpace(body.Token))
            throw new InvalidOperationException(body?.Error ?? $"Login failed ({(int)res.StatusCode})");

        _token = body.Token;
        CurrentUser = body.User;
        _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _token);
        return new AuthSession { Token = _token, User = body.User };
    }

    public async Task<UserDto?> RestoreSessionAsync(string token, CancellationToken ct = default)
    {
        _token = token;
        _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
        var res = await _http.GetAsync("/api/auth/session", ct);
        var body = await res.Content.ReadFromJsonAsync<SessionResponse>(cancellationToken: ct);
        if (!res.IsSuccessStatusCode || body?.User is null)
            return null;
        CurrentUser = body.User;
        return body.User;
    }

    public async Task LogoutAsync(CancellationToken ct = default)
    {
        try { await _http.DeleteAsync("/api/auth/session", ct); } catch { /* ignore */ }
        _token = null;
        CurrentUser = null;
        _http.DefaultRequestHeaders.Authorization = null;
    }
}
